core/services/ocr2/delegate.go

```diff
@@ -184,84 +185,129 @@ func (d *Delegate) BeforeJobCreated(spec job.Job) {
 		return nil
 	}
 	chain, err := d.chainSet.Get(big.NewInt(chainID))
 	if err != nil {
 		d.lggr.Error(err)
 		return nil
 	}
-	lp := chain.LogPoller()
 
-	var filters []string
+	return chain.LogPoller()
+}
+
+func (d *Delegate) filtersFromSpec(spec *job.OCR2OracleSpec) (filters []evmlogpoller.Filter, err error) {
 	switch spec.PluginType {
 	case job.OCR2VRF:
-		filters, err = ocr2coordinator.FilterNamesFromSpec(spec)
+		filters, err = ocr2coordinator.FiltersFromSpec(spec)
 		if err != nil {
-			d.lggr.Errorw("failed to derive ocr2vrf filter names from spec", "err", err, "spec", spec)
+			return nil, errors.Wrapf(err, "failed to derive ocr2vrf filter names from spec: %v", spec)
 		}
 	case job.OCR2Keeper:
-		filters, err = ocr2keeper.FilterNamesFromSpec(spec)
+		filters, err = ocr2keeper.FiltersFromSpec(spec)
 		if err != nil {
-			d.lggr.Errorw("failed to derive ocr2keeper filter names from spec", "err", err, "spec", spec)
+			return nil, errors.Wrapf(err, "failed to derive ocr2keeper filter names from spec: %v", spec)
 		}
 	default:
-		return nil
 	}
 
-	rargs := types.RelayArgs{
-		ExternalJobID: jb.ExternalJobID,
-		JobID:         spec.ID,
-		ContractID:    spec.ContractID,
-		New:           false,
-		RelayConfig:   spec.RelayConfig.Bytes(),
+	return filters, nil
+}
+
+func (d *Delegate) OnCreateJob(jb job.Job, q pg.Queryer) error {
+	spec := jb.OCR2OracleSpec
+	if spec == nil {
+		return ocrcommon.ErrUnexpectedJobType{Expected: job.OffchainReporting2, Actual: jb.Type}
 	}
+	if lp := d.logPollerFromSpec(spec); lp != nil {
+		filters, err := d.filtersFromSpec(jb.OCR2OracleSpec)
+		if err != nil {
+			return err
+		}
 
-	relayFilters, err := evmrelay.FilterNamesFromRelayArgs(rargs)
-	if err != nil {
-		d.lggr.Errorw("Failed to derive evm relay filter names from relay args", "err", err, "rargs", rargs)
+		for _, filter := range filters {
+			d.lggr.Debugf("Registering new filter %s", filter)
+			if err := lp.RegisterFilter(filter, q); err != nil {
+				return errors.Wrapf(err, "Failed to register filter %s", filter)
+			}
+		}
+	}
+	relayer, exists := d.relayers[spec.Relay]
+	if !exists {
+		return ocrcommon.ErrRelayNotFound{Relay: spec.Relay}
+	}
+	rargs := types.RelayArgs{
+		JobID:       spec.ID,
+		ContractID:  spec.ContractID,
+		RelayConfig: spec.RelayConfig.Bytes(),
+	}
+	r, ok := relayer.(relay.JobHooks)
+	if !ok {
 		return nil
 	}
 
-	filters = append(filters, relayFilters...)
+	return r.OnCreateJob(rargs, q)
+}
 
-	for _, filter := range filters {
-		d.lggr.Debugf("Unregistering %s filter", filter)
-		err = lp.UnregisterFilter(filter, q)
+func (d *Delegate) AfterJobCreated(spec job.Job)  {}
+func (d *Delegate) BeforeJobDeleted(spec job.Job) {}
+func (d *Delegate) OnDeleteJob(jb job.Job, q pg.Queryer) error {
+	// If the job spec is malformed in any way, we report the error but return nil so that
+	//  the job deletion itself isn't blocked.  However, if UnregisterFilter returns an
```

Comment:
why would returning an error cause blocking?

Reply:
Perhaps a poor choice of word, not blocked in the concurrency sense.  Returning an error keeps the job from being able to be deleted... because it will cause the DELETE transaction to be rolled back.  The purpose of `OnDeleteJob` is to perform any operations which need to be atomic with the job deletion so that they both succeed or fail together.